File: leetcode/p0002/Solution.java
import java.util.Arrays;
import java.util.List;

public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(-1);
        ListNode tail = dummy;
        int sum = 0;
        while (l1 != null || l2 != null) {
            sum += l1 != null ? l1.val : 0;
            sum += l2 != null ? l2.val : 0;
            int n = sum % 10;
            ListNode node = new ListNode(n);
            tail.next = node;
            tail = node;
            sum /= 10;

            if ( l1 != null ) l1 = l1.next;
            if ( l2 != null ) l2 = l2.next;
        }
        while (sum != 0) {
            int n = sum % 10;
            ListNode node = new ListNode(n);
            tail.next = node;
            tail= node;
            sum /= 10;
        }
        return dummy.next;
    }

    public static void main(String[] args) {
        List<Input> inputs = Arrays.asList(new Input[]{new Input(342, 465)});
        for (Input input : inputs) {
            ListNode num1 = makeList(input.lhs);
            ListNode num2 = makeList(input.rhs);
            System.out.printf(" Input:  %s + %s\n", num1, num2);

            Solution solution = new Solution();
            ListNode result = solution.addTwoNumbers(num1, num2);
            System.out.printf(" Output: %s\n", result);
        }
    }

    private static ListNode makeList(int val) {
        if (val < 0) return null;
        ListNode dummy = new ListNode(-1);
        ListNode head = dummy;
        while (val != 0) {
            int n = val % 10;
            ListNode node = new ListNode(n);
            head.next = node;
            head = node;
            val /= 10;
        }
        return dummy.next;
    }

    static final class Input {
        private int lhs;
        private int rhs;

        Input(int lhs, int rhs) {
            this.lhs = lhs;
            this.rhs = rhs;
        }
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) { val = x; }

    @Override
    public String toString() {
        StringBuilder buf = new StringBuilder("(");
        ListNode node = this;
        while (node != null) {
            if (node != this) {
                buf.append(" -> ");
            }
            buf.append(node.val);
            node = node.next;
        }
        buf.append(")");
        return buf.toString();
    }
}

File: leetcode/p0002/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct Object {
    char *lhs;
    char *rhs;
};

struct ListNode {
    int val;
    struct ListNode *next;
};

void print_num(struct ListNode *node)
{
    if (node == NULL) {
        return;
    }
    printf("(%d", node->val);
    while ((node = node->next) != NULL) {
        printf(" -> %d", node->val);
    }
    printf(")");
}

struct ListNode *make_num(const char *digits)
{
    int len = strlen(digits);
    const char *c = digits + len - 1;
    struct ListNode dummy, *prev;

    dummy.next = NULL;
    prev = &dummy;
    while (len-- > 0) {
        struct ListNode *p = malloc(sizeof(*p));
        p->val = *c-- - '0';
        p->next = NULL;
        prev->next = p;
        prev = p;
    }

    return dummy.next;
}

void free_num(struct ListNode *node) {
    struct ListNode *n;
    while (node != NULL) {
        n = node;
        node = node->next;
        free(n);
    }
}

struct ListNode* add_two_numbers(struct ListNode* l1, struct ListNode* l2)
{
    int carry = 0;
    struct ListNode dummy, *prev;

    dummy.next = NULL;
    prev = &dummy;
    while (l1 != NULL || l2 != NULL || carry) {
        int sum = 0;

        if (l1 != NULL) {
            sum += l1->val;
            l1 = l1->next;
        }
        if (l2 != NULL) {
            sum += l2->val;
            l2 = l2->next;
        }
        sum += carry;

        struct ListNode *p = malloc(sizeof(*p));
        p->val = sum % 10;
        p->next = NULL;
        prev->next = p;
        prev = p;

        carry = sum / 10;
    }

    return dummy.next;
}

int main(int argc, char **argv)
{
    struct Object inputs[] = {
        { .lhs = "342", .rhs = "465" },
    };
    int i, len = sizeof(inputs) / sizeof(struct Object);

    for (i = 0; i < len; i++) {
        struct ListNode *lhs = make_num(inputs[i].lhs);
        struct ListNode *rhs = make_num(inputs[i].rhs);
        struct ListNode *sum = add_two_numbers(lhs, rhs);
        int n1 = atoi(inputs[i].lhs);
        int n2 = atoi(inputs[i].rhs);

        printf("\n Input: ");
        print_num(lhs);
        printf(" + ");
        print_num(rhs);
        printf("\n Output: ");
        print_num(sum);

        printf("\n Explanation: %d + %d = %d\n", n1, n2, n1 + n2);

        free_num(lhs);
        free_num(rhs);
        free_num(sum);
    }

    return EXIT_SUCCESS;
}
